core/server/worker/src/main/java/alluxio/worker/grpc/AbstractReadHandler.java

```diff
@@ -126,14 +152,36 @@ public void onNext(alluxio.grpc.ReadRequest request) {
       mContext.setPosReceived(mContext.getRequest().getStart());
       mDataReaderExecutor.submit(createDataReader(mContext, mResponseObserver));
       mContext.setDataReaderActive(true);
+    } catch (RejectedExecutionException e) {
+      handleStreamEndingException(e, Status.RESOURCE_EXHAUSTED);
     } catch (Exception e) {
-      LogUtils.warnWithException(LOG, "Exception occurred while processing read request {}.",
-          request, e);
-      mSerializingExecutor.execute(() -> mResponseObserver
-          .onError(AlluxioStatusException.fromCheckedException(e).toGrpcStatusException()));
+      handleStreamEndingException(
+          AlluxioStatusException.fromThrowable(e).toGrpcStatusException());
     }
   }
 
+  private void handleStreamEndingException(StatusException e) {
+    handleStreamEndingException(e, e.getStatus());
+  }
+
+  // In this case, we hit an exception that does not allow us to proceed further
+  private void handleStreamEndingException(Exception e, Status status) {
+    // The executor is too busy to handle our requests. The stream should error.
+    Long sessionId = mContext.getRequest() == null ? null : mContext.getRequest().getSessionId();
+    LogUtils.warnWithException(LOG, "handling RejectedExecutionException: sessionId: {}",
+        sessionId, e);
+    AlluxioStatusException statusExc =
+        new AlluxioStatusException(status);
+    if (mContext == null) {
+      mContext = createRequestContext(alluxio.grpc.ReadRequest.newBuilder().build());
+    }
+    setError(new Error(statusExc, true));
+    // After calling setError, this will run the data reader to complete the request and close
+    // any possibly opened blocks, then replies to the client with onError, and should eventually
+    // shut down the serializing executor.
+    createDataReader(mContext, mResponseObserver).run();
```

Comment:
Previously we do all the work in a separate thread ensuring abstract read handler methods execute fairly quickly. Will executing the data reader on this thread cause significant slowdowns?

Reply:
After calling `setError`, the context status will have the `mError` flag as true. Once the error flag is set calling `createDataReader(...).run()` should be relatively fast.

There are a few places where we could see a slowdown.
- Entering the loop where we wait for `mLock`
  - https://github.com/Alluxio/alluxio/blob/557484bf6cf6923dc55ca7bfe63297dab02300c7/core/server/worker/src/main/java/alluxio/worker/grpc/AbstractReadHandler.java#L358-L375
  - Because we call `setError` before running the DataReader, we immediately break from the while loop, as lock as we can access `mLock`.
- Afterwards we need to then perform any unlockBlock operations via the `completeRequest`. This should complete quickly.
  - https://github.com/Alluxio/alluxio/blob/557484bf6cf6923dc55ca7bfe63297dab02300c7/core/server/worker/src/main/java/alluxio/worker/grpc/AbstractReadHandler.java#L431-L456
- Then finally will call `replyError`, which prevents any more operations from occurring. It will shutdown the serializing executor as well.
  - https://github.com/Alluxio/alluxio/blob/557484bf6cf6923dc55ca7bfe63297dab02300c7/core/server/worker/src/main/java/alluxio/worker/grpc/AbstractReadHandler.java#L480-L491

So yes, I think this does have some potential to slow down the handler threads.

I haven't tested yet, but I don't think calling `setError` is enough to guarantee that we call `replyError`. The problem arises if there aren't enough executor threads to actually create the new data reader to send the final `mResponseObserve.onError` request. Running it here is a proactive measure against another possible `RejectedExecutionException`

Reply:
Maybe instead we can make a comprise. As long as we submit the data reader to the executor successfully, we don't really care when it executes. Just that it does get run at some point after `setError`.

So in the handler here, we can do

```java
try {
  mDataReaderExecutor.submit(createDataReader(...));
} catch (RejectedExecutionException e) {
  createDataReader(...).run();
}
```

What do you think?

Reply:
How about a separate threadpool with infinite size queue?

Reply:
Right now we limit the number of clients that can read from a single worker by the max size of threadpool that we submit data readers to. By making that queue size unlimited, it makes it possible for the queue to grow excessively large under high load and the latency between submitting a data reader and actually beginning to process could end up being larger than client timeouts. Right now, when we submit a data reader, it's guaranteed to run immediately unless the thread pool is at max capacity.

I think it's better to fail fast when reaching a specified load on the worker, rather than letting  too many clients connect. With the new retries it will force the clients to try a new worker quicker than if we use an infinite-size queue.

Reply:
This code has been removed